### ISSUE TITLE
Spawn and pipe exception

### ISSUE THREAD

**Issue by orodeh** [joyent/node#15518]
We have a javascript program that uses nodejs. It spawns a process and communicates with it using pipes. The child process encounters a problem, and exists, the result is not caught by any standard method. We get an exception that looks like this:

events.js:72
        throw er; // Unhandled 'error' event
              ^

Error: read ECONNRESET
    at errnoException (net.js:904:11)
    at Pipe.onread (net.js:558:19)

This is despite the existence of code to handle error conditions. Here is a sketch of the code: 

// note: data is a string of limited length
f(data) { 
   var scanProc = childProcess.spawn('clamdscan', ['--quiet', '-'], {cwd: '/tmp', stdio: 'pipe'});
 
   // error/signal handling code
   scanProc.on('exit', function(code, signal) { ... }
   scanProc.on('error', function(err) { .. }
 
   scanProc.stdin.write(data);
   scanProc.stdin.end();
}

In other words, the exception is thrown in an underlying system thread, and it is not propagated to the thread that called the spawn operation, or its callbacks. 

Environment:
   Ubuntu 12.04
   nodejs version: v0.10.32

**Comment by cjihrig** [comment]
@orodeh could you please provide a reduced, runnable example?

**Comment by piscisaureus** [comment]
You should add an error handler to the stream that throws the error, not just the ChildProcess.object. So:

```js
scanProc.stdout.on('error', function(err) { ... });
scanProc.stderr.on('error', function(err) { ... });
```

**Comment by orodeh** [comment]
Thank you! That fixed the problem.

      Ohad.


On Wed, Apr 15, 2015 at 3:17 PM, Bert Belder <notifications@github.com>
wrote:

> You should add an error handler to the stream that throws the error, not
> just the ChildProcess.object. So:
>
> scanProc.stdout.on('error', function(err) { ... });
> scanProc.stderr.on('error', function(err) { ... });
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/issues/15518#issuecomment-93584925>.
>

-- 
---
The contents of this e-mail and any attachments are confidential and only 
for use by the intended recipient. Any unauthorized use, distribution or 
copying of this message is strictly prohibited. If you are not the intended 
recipient please inform the sender immediately by reply e-mail and delete 
this message from your system. Thank you for your co-operation.
